src/services/signer_service.ts

```diff
@@ -0,0 +1,183 @@
+        return [...Array(rangeCount).keys()];
+    }
+    private static _calculateProtocolFee(numOrders: number, gasPrice: BigNumber): BigNumber {
+        return new BigNumber(150000).times(gasPrice).times(numOrders);
+    }
+    constructor() {
+        this._privateWalletSubprovider = new PrivateKeyWalletSubprovider(META_TXN_RELAY_PRIVATE_KEY);
+        this._nonceTrackerSubprovider = new NonceTrackerSubprovider();
+        this._provider = SignerService._createWeb3Provider(
+            ETHEREUM_RPC_URL,
+            this._privateWalletSubprovider,
+            this._nonceTrackerSubprovider,
+        );
+        this._contractWrappers = new ContractWrappers(this._provider, { chainId: CHAIN_ID });
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+    }
+    public async validateZeroExTransactionFillAsync(
+        zeroExTransaction: ZeroExTransactionWithoutDomain,
+        signature: string,
+    ): Promise<BigNumber> {
+        // Verify 0x txn won't expire in next 60 seconds
+        // tslint:disable-next-line:custom-no-magic-numbers
+        const sixtySecondsFromNow = new BigNumber(+new Date() + 60);
+        if (zeroExTransaction.expirationTimeSeconds <= sixtySecondsFromNow) {
+            throw new Error('zeroExTransaction expirationTimeSeconds in less than 60 seconds from now');
+        }
+
+        const devUtils = new DevUtilsContract(this._contractWrappers.contractAddresses.devUtils, this._provider);
+        const decodedArray = await devUtils.decodeZeroExTransactionData(zeroExTransaction.data).callAsync();
+        const orders = decodedArray[1];
+
+        // Verify orders don't expire in next 60 seconds
+        orders.forEach(order => {
+            if (order.expirationTimeSeconds <= sixtySecondsFromNow) {
+                throw new Error('Order included in zeroExTransaction expires in less than 60 seconds from now');
+            }
+        });
+
+        const gasPrice = zeroExTransaction.gasPrice;
+        const protocolFee = SignerService._calculateProtocolFee(orders.length, gasPrice);
+
+        try {
+            await this._contractWrappers.exchange.executeTransaction(zeroExTransaction, signature).callAsync({
+                from: META_TXN_RELAY_ADDRESS,
+                gasPrice,
+                value: protocolFee,
+            });
+        } catch (err) {
+            if (err.values && err.values.errorData && err.values.errorData !== '0x') {
+                const decodedCallData = RevertError.decode(err.values.errorData, false);
+                throw decodedCallData;
+            }
+            throw err;
+        }
+
+        return protocolFee;
+    }
+    public async generateExecuteTransactionEthereumTransactionAsync(
+        zeroExTransaction: ZeroExTransactionWithoutDomain,
+        signature: string,
+        protocolFee: BigNumber,
+    ): Promise<PartialTxParams> {
+        const gasPrice = zeroExTransaction.gasPrice;
+        const gas = await this._contractWrappers.exchange
```

Comment:
TODO our pattern is to `eth_call` and `estimateGas` in parallel and return the result of `eth_call` validations

Reply:
Added a TODO.